km/km_filesys.c

```diff
@@ -1951,27 +1962,55 @@ static inline void do_full_write(int fd, char* bufp, size_t bufl)
 static inline size_t
 fs_core_save_pipe_contents(char* buf, size_t length, km_file_t* file, int writefd, size_t queuedbytes)
 {
+   int fd = km_filet2fd(file);
    km_assert(queuedbytes <= length);
+   km_assert(fd >= 0);
 
-   // Read the bytes queued in the pipe
-   do_full_read(file->ofd, buf, queuedbytes);
+   // Read the bytes queued in the pipe into the memory buf points to
+   do_full_read(fd, buf, queuedbytes);
 
-   // now write the bytes back into the pipe
-   do_full_write(writefd, buf, queuedbytes);
+   if (writefd < 0) {
+      // The write side of this pipe is closed.  So we need to recreate the pipe
+      // to put the data back into the pipe buffer and then close the write end
+      // after writing the data.
+      close(fd);
```

Comment:
as you mentioned. You may want to remove this line.
If nothing else it makes code easier to analyse wrt same fd being given back on pipe2/sockerpair.

Reply:
(i.e) you got a fd same as this and you dup and close the same fd.

Reply:
Will fix.